### ISSUE TITLE
LateinitUsage should support ignoring certain types

### ISSUE THREAD

**Issue by 3flex** [detekt/detekt#3053]
## Expected Behavior of the rule
I have this line:
```
private lateinit var binding: ActivityRecordBinding
```
Which should not be flagged by LateinitUsage.

## Context
It's not possible to set `binding` to any value until after the class is instantiated. Because I am using View Binding on Android, and `ActivityRecordBinding` extends `ViewBinding`, it would be nice to be able to exclude all lateinit usage for anything that extends `androidx.viewbinding.ViewBinding`. Similarly it would be great to see this for anything that extends `androidx.databinding.ViewDataBinding`.

**Comment by BraisGabin** [comment]
I have this rule disabled for things like this. But I'm not sure if the solution is to check the type of the lateinit. What about filter this rule in all the clases that extends activity/fragment? You can't manage what happen in the construction of those classes so lateinit have sense there.

**Comment by 3flex** [comment]
I see your point, but in Android `lateinit` should generally only needed for dependency injection or when a variable references a view, activity or fragment. Viewbinding/databinding is a common use case. It should generally not be used at all for cases that don't meet those criteria.

If the rule can ignore certain types, that will allow fewer false positives identified by the rule and make it more useful on Android instead of disabling it completely or disabling it completely for fragments and activities. It's similar in nature to the existing ability to ignore variables with particular annotations.

**Comment by schalkms** [comment]
Okay, so what is the proposal here. Do you want to add a new configuration attribute, which specifies the type being ignored?

**Comment by 3flex** [comment]
Yes, exactly

**Comment by cortinico** [comment]
> Do you want to add a new configuration attribute, which specifies the type being ignored?
> Yes, exactly

Isn't the original issue related to subclass types?

> it would be nice to be able to exclude all lateinit usage for anything that extends androidx.viewbinding.ViewBinding. Similarly it would be great to see this for anything that extends androidx.databinding.ViewDataBinding.

Imho if we filter for property types, than it probably makes sense.
If we filter for types of the superclass, that we're designing this rule around an implementation assumption and I would avoid it.

**Comment by 3flex** [comment]
> Isn't the original issue related to subclass types?

I'd want both options. I want to whitelist anything that extends `androidx.viewbinding.ViewBinding`, instead of having to whitelist every single viewbinding in the app - there could be one for every fragment and activity in the app and having to add every single one to the detekt config gets annoying.

I'd also want to whitelist specific types only.

> if we filter for property types, than it probably makes sense.
Hopefully we're thinking the same thing then. Basically if I have this:
```kotlin
class RecordEditFragment : Fragment() {
    private lateinit var binding: FragmentRecordEditBinding
    // ...
}
```
Right now, that `lateinit` usage is reported. To avoid that, since `lateinit` is appropriate here, I would want to be able to add to the config either `com.app.databinding.FragmentRecordEditBinding` or `androidx.databinding.ViewDataBinding` so it's not reported. I'd use the second if I just want to whitelist any view binding, and the first if I'm dealing with a specific type (maybe I'm injecting a database class with Hilt into a class that doesn't support Hilt - in that case I can't whitelist `@Inject` because the annotation wouldn't be used, so I'd whitelist the specific type `com.app.database.Database`).

I'm just trying to think of a way to make this rule still have value on Android, because that's where `lateinit` is frequently needed. Supporting whitelisting of certain types solves for the examples I provided above which are common on Android.

**Comment by BraisGabin** [comment]
I know, this is just an example but this code is not good:

```kotlin
class RecordEditFragment : Fragment() {
    private lateinit var binding: FragmentRecordEditBinding
    // ...
}
```

You are leaking the view after the `onViewDestroyed`. So, even the use of `lateinit` for the type `androidx.databinding.ViewDataBinding` should not be allowed worldwide. It's only correct in some contexts as the Activities.

I'm not against adding this configuration, but this seems like a real complex and specific topic. 🤔 could it have sense to create a new repo under the @detekt organization for android rules? Or should it be a third party one?

**Comment by schalkms** [comment]
Maybe there should be an Android ruleset.
